Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -119,29 +121,48 @@ public bool TryGetXComponentCorrelationId(string instrumentationKey, out string
-                    if (getAppIdTask.Wait(GetAppIdTimeout))
-                    {
-                        this.GenerateCorrelationIdAndAddToDictionary(instrumentationKey, getAppIdTask.Result);
-                        correlationId = this.knownCorrelationIds[instrumentationKey];
-                        return true;
-                    }
-                    else
+                    // We only want one task to be there to fetch the ikey. If initial requests come in a bunch, only one of them gets to take responsibility of creating the fetch task. Rest return.
+                    if (this.FetchTasks.TryAdd(instrumentationKey, int.MinValue))
                     {
-                        getAppIdTask.ContinueWith((appId) =>
+                        try
                         {
-                            try
+                            // We wait for <getAppIdTimeout> seconds (which is 0 at this point) to retrieve the appId. If retrieved during that time, we return success setting the correlation id.
+                            // If we are still waiting on the result beyond the timeout - for this particular call we return the failure but queue a task continuation for it to be cached for next time.
+                            Task<string> getAppIdTask = this.provideAppId(instrumentationKey.ToLowerInvariant());
+
+                            if (getAppIdTask.Wait(GetAppIdTimeout))
                             {
-                                this.GenerateCorrelationIdAndAddToDictionary(instrumentationKey, appId.Result);
+                                this.GenerateCorrelationIdAndAddToDictionary(instrumentationKey, getAppIdTask.Result);
+                                correlationId = this.knownCorrelationIds[instrumentationKey];
+                                return true;
                             }
-                            catch (Exception ex)
+                            else
                             {
-                                this.RegisterFailure(instrumentationKey, ex);
+                                getAppIdTask.ContinueWith((appId) =>
+                                {
+                                    try
+                                    {
+                                        this.GenerateCorrelationIdAndAddToDictionary(instrumentationKey, appId.Result);
+                                    }
+                                    catch (Exception ex)
+                                    {
+                                        this.RegisterFailure(instrumentationKey, ex);
+                                    }
+                                });
+
+                                return false;
                             }
-                        });
-
+                        }
+                        finally
+                        {
+                            int taskId;
+                            this.FetchTasks.TryRemove(instrumentationKey, out taskId);
```

Comment:
good point - that is definitely an issue. the finally block should be in the continuation - my bad.